File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/components/containers/SpriteContainerView.kt
package com.crescentflare.slicinggame.components.containers

import android.annotation.TargetApi
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PointF
import android.os.Build
import android.util.AttributeSet
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.components.utility.ViewletUtil
import com.crescentflare.slicinggame.infrastructure.geometry.Polygon
import com.crescentflare.slicinggame.infrastructure.geometry.Vector
import com.crescentflare.slicinggame.infrastructure.physics.Physics
import com.crescentflare.slicinggame.infrastructure.physics.PhysicsBoundary
import com.crescentflare.slicinggame.sprites.core.Sprite
import com.crescentflare.slicinggame.sprites.core.SpriteCanvas
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.lang.ref.WeakReference
import kotlin.math.PI
import kotlin.math.atan2
import kotlin.math.max


/**
 * Container view: provides a container for managing sprites
 */
open class SpriteContainerView : FrameContainerView, Physics.Listener {

    // --
    // Static: viewlet integration
    // --

    companion object {

        val viewlet: JsonInflatable = object : JsonInflatable {
            override fun create(context: Context): Any {
                return SpriteContainerView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is SpriteContainerView) {
                    // Apply grid size
                    obj.gridWidth = mapUtil.optionalFloat(attributes, "gridWidth", 1f)
                    obj.gridHeight = mapUtil.optionalFloat(attributes, "gridHeight", 1f)
                    obj.sliceWidth = mapUtil.optionalFloat(attributes, "sliceWidth", 0f)

                    // Apply update frames per second
                    obj.fps = mapUtil.optionalInteger(attributes, "fps", 60)

                    // Apply debug settings
                    obj.drawPhysicsBoundaries = mapUtil.optionalBoolean(attributes, "drawPhysicsBoundaries", false)

                    // Apply sprites
                    val spriteList = mapUtil.optionalObjectList(attributes, "sprites")
                    obj.clearSprites()
                    for (spriteItem in spriteList) {
                        mapUtil.asStringObjectMap(spriteItem)?.let {
                            val sprite = Sprite()
                            sprite.x = mapUtil.optionalFloat(it, "x", 0f)
                            sprite.y = mapUtil.optionalFloat(it, "y", 0f)
                            sprite.width = mapUtil.optionalFloat(it, "width", 1f)
                            sprite.height = mapUtil.optionalFloat(it, "height", 1f)
                            obj.addSprite(sprite)
                        }
                    }

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj::class == SpriteContainerView::class
            }
        }
    }


    // --
    // Members
    // --

    var physicsListener: Physics.Listener?
        get() = physicsListenerReference?.get()
        set(physicsListener) {
            physicsListenerReference = if (physicsListener != null) {
                WeakReference(physicsListener)
            } else {
                null
            }
        }

    private val physics = Physics()
    private val sprites = mutableListOf<Sprite>()
    private var collisionBoundaries = mutableListOf<PhysicsBoundary>()
    private var sliceVectorBoundary: PhysicsBoundary? = null
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val spriteCanvas = SpriteCanvas(paint)
    private var currentSliceVector: Vector? = null
    private var updateScheduled = false
    private var lastTimeMillis = System.currentTimeMillis()
    private var timeCorrection = 1
    private var physicsListenerReference: WeakReference<Physics.Listener>? = null


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        setWillNotDraw(false)
        physics.listener = this
    }


    // --
    // Sprites
    // --

    fun addSprite(sprite: Sprite) {
        sprites.add(sprite)
        physics.registerObject(sprite)
    }

    fun clearSprites() {
        for (sprite in sprites) {
            physics.unregisterObject(sprite)
        }
        sprites.clear()
    }


    // --
    // Collision boundaries
    // --

    fun addCollisionBoundary(boundary: PhysicsBoundary) {
        collisionBoundaries.add(boundary)
        physics.registerObject(boundary)
    }

    fun addCollisionBoundaries(polygon: Polygon) {
        val boundaryWidth = max(sliceWidth, gridWidth * 0.005f)
        for (vector in polygon.asVectorList()) {
            val offsetVector = vector.perpendicular().unit() * (boundaryWidth / 2)
            val halfDistanceX = vector.x / 2
            val halfDistanceY = vector.y / 2
            val vectorCenterX = vector.start.x + halfDistanceX
            val vectorCenterY = vector.start.y + halfDistanceY
            val centerX = vectorCenterX + offsetVector.x
            val centerY = vectorCenterY + offsetVector.y
            val vectorLength = vector.distance()
            val x = centerX - boundaryWidth / 2
            val y = centerY - vectorLength / 2
            val rotation = atan2(vector.x, vector.y) * 360 / (PI.toFloat() * 2)
            addCollisionBoundary(PhysicsBoundary(x, y, boundaryWidth, vectorLength, -rotation))
        }
    }

    fun clearCollisionBoundaries() {
        for (collisionBoundary in collisionBoundaries) {
            physics.unregisterObject(collisionBoundary)
        }
        collisionBoundaries.clear()
    }


    // --
    // Slicing
    // --

    fun setSliceVector(vector: Vector?): Boolean {
        // First unregister the existing object
        val previousSliceVector = currentSliceVector
        sliceVectorBoundary?.let {
            physics.unregisterObject(it)
        }
        sliceVectorBoundary = null
        currentSliceVector = null

        // Check if it already collides
        vector?.let {
            if (previousSliceVector != null) {
                val intersection = previousSliceVector.intersect(it)
                val polygons = mutableListOf<Polygon>()
                if (intersection != null) {
                    polygons.add(Polygon(listOf(previousSliceVector.start, vector.start, intersection)))
                    polygons.add(Polygon(listOf(previousSliceVector.end, vector.end, intersection)))
                } else {
                    polygons.add(Polygon(listOf(previousSliceVector.start, vector.start, vector.end, previousSliceVector.end)))
                }
                for (polygon in polygons) {
                    val checkPolygon = if (polygon.isClockwise()) polygon else polygon.reversed()
                    if (physics.intersectsSprite(checkPolygon)) {
                        onLethalCollision()
                        return false
                    }
                }
            } else if (physics.intersectsSprite(it)) {
                onLethalCollision()
                return false
            }
        }

        // Add slice boundary
        vector?.let {
            val vectorCenterX = it.start.x + it.x / 2
            val vectorCenterY = it.start.y + it.y / 2
            val width = gridWidth * 0.005f
            val height = it.distance()
            val x = vectorCenterX - width / 2
            val y = vectorCenterY - height / 2
            val rotation = atan2(it.x, it.y) * 360 / (PI.toFloat() * 2)
            val physicsBoundary = PhysicsBoundary(x, y, width, height, -rotation)
            physicsBoundary.lethal = true
            physics.registerObject(physicsBoundary)
            sliceVectorBoundary = physicsBoundary
        }
        currentSliceVector = vector
        return true
    }

    fun spritesOnPolygon(polygon: Polygon): Boolean {
        return physics.intersectsSprite(polygon)
    }

    fun spritesPerSlice(vector: Vector, polygon: Polygon): Int {
        var spriteCount = 0
        sprites.forEach {
            val spriteBounds = it.collisionBounds
            spriteBounds.offset(it.x, it.y)
            val spritePolygon = Polygon(spriteBounds, PointF(it.collisionPivot.x + it.x, it.collisionPivot.y + it.y), it.collisionRotation)
            if (spritePolygon.intersect(polygon)) {
                if (vector.directionOfPoint(PointF(it.x + it.collisionPivot.x, it.y + it.collisionPivot.y)) >= 0) {
                    spriteCount += 1
                }
            }
        }
        return spriteCount
    }


    // --
    // Configurable values
    // --

    var gridWidth: Float = 1f
        set(gridWidth) {
            val changed = field != gridWidth
            field = gridWidth
            if (changed) {
                requestLayout()
            }
            physics.width = gridWidth
        }

    var gridHeight: Float = 1f
        set(gridHeight) {
            val changed = field != gridHeight
            field = gridHeight
            if (changed) {
                requestLayout()
            }
            physics.height = gridHeight
        }

    var sliceWidth = 0f

    var fps = 60

    var drawPhysicsBoundaries = false


    // --
    // Physics listener
    // --

    override fun onLethalCollision() {
        physicsListener?.onLethalCollision()
    }


    // --
    // Movement
    // --

    private fun update(timeDifference: Long) {
        val timeInterval = timeDifference.toFloat() / 1000
        physics.prepareObjects()
        sprites.forEach { sprite ->
            sprite.update(timeInterval, physics)
        }
        invalidate()
    }


    // --
    // Drawing
    // --

    override fun onDraw(canvas: Canvas?) {
        // Draw sprites and optional physics boundaries
        canvas?.let {
            it.clipRect(0, 0, width, height)
            spriteCanvas.prepare(it, width.toFloat(), height.toFloat(), gridWidth, gridHeight)
            sprites.forEach { sprite ->
                sprite.draw(spriteCanvas)
            }
            if (drawPhysicsBoundaries) {
                for (boundary in collisionBoundaries) {
                    spriteCanvas.fillRotatedRect(boundary.x + boundary.collisionPivot.x, boundary.y + boundary.collisionPivot.y, boundary.width, boundary.height, Color.RED, boundary.collisionRotation)
                }
                sliceVectorBoundary?.let { boundary ->
                    spriteCanvas.fillRotatedRect(boundary.x + boundary.collisionPivot.x, boundary.y + boundary.collisionPivot.y, boundary.width, boundary.height, Color.YELLOW, boundary.collisionRotation)
                }
            }
        }

        // Schedule next update
        if (!updateScheduled) {
            val checkTimeMillis = System.currentTimeMillis()
            val delayTime = 1000 / fps - (checkTimeMillis - lastTimeMillis)
            updateScheduled = true
            GlobalScope.launch(Dispatchers.Main) {
                // Delay and try to correct for time lost due to coroutine inaccuracy
                delay(max(1, delayTime - timeCorrection))
                val currentTimeMillis = System.currentTimeMillis()
                if (delayTime >= 1) {
                    val lostTimeMillis = (currentTimeMillis - checkTimeMillis) - delayTime
                    if (lostTimeMillis < -1) {
                        timeCorrection -= 1
                    } else if (lostTimeMillis > 1) {
                        timeCorrection += 1
                    }
                }

                // Continue with the next update
                var difference = currentTimeMillis - lastTimeMillis
                lastTimeMillis = currentTimeMillis
                updateScheduled = false
                if (difference > 1000 / fps * 5) {
                    difference = (1000 / fps * 5).toLong()
                }
                update(difference)
            }
        }
    }


    // --
    // Custom layout
    // --

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val widthMode = MeasureSpec.getMode(widthMeasureSpec)
        val heightMode = MeasureSpec.getMode(heightMeasureSpec)
        val width = MeasureSpec.getSize(widthMeasureSpec)
        val height = MeasureSpec.getSize(heightMeasureSpec)
        if (widthMode == MeasureSpec.AT_MOST && heightMode == MeasureSpec.AT_MOST) {
            if (width * gridHeight / gridWidth <= height) {
                setMeasuredDimension(width, (width * gridHeight / gridWidth).toInt())
            } else {
                setMeasuredDimension((height * gridWidth / gridHeight).toInt(), height)
            }
        } else if (widthMode == MeasureSpec.EXACTLY && heightMode == MeasureSpec.EXACTLY) {
            setMeasuredDimension(width, height)
        } else if (widthMode == MeasureSpec.AT_MOST || widthMode == MeasureSpec.EXACTLY) {
            setMeasuredDimension(width, (width * gridHeight / gridWidth).toInt())
        } else if (heightMode == MeasureSpec.AT_MOST || heightMode == MeasureSpec.EXACTLY) {
            setMeasuredDimension((height * gridWidth / gridHeight).toInt(), height)
        } else {
            super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        }
    }

}

File: slicing-ios/SlicingGame/Components/Containers/SpriteContainerView.swift
//
//  SpriteContainerView.swift
//  Container view: provides a container for managing sprites
//

import UIKit
import UniLayout
import JsonInflator

class SpriteContainerView: FrameContainerView, PhysicsDelegate {
    
    // --
    // MARK: Members
    // --
    
    weak var physicsDelegate: PhysicsDelegate?
    private let physics = Physics()
    private var sprites = [Sprite]()
    private var collisionBoundaries = [PhysicsBoundary]()
    private var sliceVectorBoundary: PhysicsBoundary?
    private var currentSliceVector: Vector?
    private var updateScheduled = false
    private var lastTimeInterval = Date.timeIntervalSinceReferenceDate
    private var timeCorrection = 0.001


    // --
    // MARK: Viewlet integration
    // --

    override class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return SpriteContainerView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let spriteContainer = object as? SpriteContainerView {
                // Apply grid size
                spriteContainer.gridWidth = convUtil.asFloat(value: attributes["gridWidth"]) ?? 1
                spriteContainer.gridHeight = convUtil.asFloat(value: attributes["gridHeight"]) ?? 1
                spriteContainer.sliceWidth = convUtil.asFloat(value: attributes["sliceWidth"]) ?? 0

                // Apply update frames per second
                spriteContainer.fps = convUtil.asInt(value: attributes["fps"]) ?? 60

                // Apply debug settings
                spriteContainer.drawPhysicsBoundaries = convUtil.asBool(value: attributes["drawPhysicsBoundaries"]) ?? false

                // Apply sprites
                spriteContainer.clearSprites()
                if let spriteList = attributes["sprites"] as? [[String: Any]] {
                    for spriteItem in spriteList {
                        let sprite = Sprite()
                        sprite.x = convUtil.asFloat(value: spriteItem["x"]) ?? 0
                        sprite.y = convUtil.asFloat(value: spriteItem["y"]) ?? 0
                        sprite.width = convUtil.asFloat(value: spriteItem["width"]) ?? 1
                        sprite.height = convUtil.asFloat(value: spriteItem["height"]) ?? 1
                        spriteContainer.addSprite(sprite)
                    }
                }

                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: spriteContainer, attributes: attributes)
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return type(of: object) == SpriteContainerView.self
        }
        
    }
    
    
    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    private func setup() {
        physics.delegate = self
    }
    
    
    // --
    // MARK: Sprites
    // --
    
    func addSprite(_ sprite: Sprite) {
        sprites.append(sprite)
        physics.registerObject(sprite)
    }
    
    func clearSprites() {
        for sprite in sprites {
            physics.unregisterObject(sprite)
        }
        sprites.removeAll()
    }


    // --
    // MARK: Collision boundaries
    // --
    
    func addCollisionBoundary(_ boundary: PhysicsBoundary) {
        collisionBoundaries.append(boundary)
        physics.registerObject(boundary)
    }
    
    func addCollisionBoundaries(fromPolygon: Polygon) {
        let boundaryWidth = max(sliceWidth, gridWidth * 0.005)
        for vector in fromPolygon.asVectorArray() {
            let offsetVector = vector.perpendicular().unit() * CGFloat(boundaryWidth / 2)
            let halfDistanceX = vector.x / 2
            let halfDistanceY = vector.y / 2
            let vectorCenterX = Float(vector.start.x + halfDistanceX)
            let vectorCenterY = Float(vector.start.y + halfDistanceY)
            let centerX = vectorCenterX + Float(offsetVector.x)
            let centerY = vectorCenterY + Float(offsetVector.y)
            let vectorLength = Float(vector.distance())
            let x = centerX - boundaryWidth / 2
            let y = centerY - vectorLength / 2
            let rotation = atan2(vector.x, vector.y) * 360 / (CGFloat.pi * 2)
            addCollisionBoundary(PhysicsBoundary(x: x, y: y, width: boundaryWidth, height: vectorLength, rotation: Float(-rotation)))
        }
    }
    
    func clearCollisionBoundaries() {
        for collisionBoundary in collisionBoundaries {
            physics.unregisterObject(collisionBoundary)
        }
        collisionBoundaries.removeAll()
    }


    // --
    // MARK: Slicing
    // --

    func setSliceVector(vector: Vector?) -> Bool {
        // First unregister the existing object
        let previousSliceVector = currentSliceVector
        if let sliceVectorBoundary = sliceVectorBoundary {
            physics.unregisterObject(sliceVectorBoundary)
        }
        sliceVectorBoundary = nil
        currentSliceVector = nil

        // Check if it already collides
        if let vector = vector {
            if let previousVector = previousSliceVector {
                var polygons = [Polygon]()
                if let intersection = previousVector.intersect(withVector: vector) {
                    polygons.append(Polygon(points: [ previousVector.start, vector.start, intersection ]))
                    polygons.append(Polygon(points: [ previousVector.end, vector.end, intersection ]))
                } else {
                    polygons.append(Polygon(points: [ previousVector.start, vector.start, vector.end, previousVector.end ]))
                }
                for polygon in polygons {
                    let checkPolygon = polygon.isClockwise() ? polygon : polygon.reversed()
                    if physics.intersectsSprite(polygon: checkPolygon) {
                        didLethalCollision()
                        return false
                    }
                }
            } else if physics.intersectsSprite(vector: vector) {
                didLethalCollision()
                return false
            }
        }

        // Add slice boundary
        if let vector = vector {
            let vectorCenterX = vector.start.x + vector.x / 2
            let vectorCenterY = vector.start.y + vector.y / 2
            let width: CGFloat = CGFloat(gridWidth) * 0.005
            let height = vector.distance()
            let x = Float(vectorCenterX - width / 2)
            let y = Float(vectorCenterY - height / 2)
            let rotation = atan2(vector.x, vector.y) * 360 / (CGFloat.pi * 2)
            let physicsBoundary = PhysicsBoundary(x: x, y: y, width: Float(width), height: Float(height), rotation: Float(-rotation))
            physicsBoundary.lethal = true
            physics.registerObject(physicsBoundary)
            sliceVectorBoundary = physicsBoundary
        }
        currentSliceVector = vector
        return true
    }
    
    func spritesOnPolygon(polygon: Polygon) -> Bool {
        return physics.intersectsSprite(polygon: polygon)
    }
    
    func spritesPerSlice(vector: Vector, inPolygon: Polygon) -> Int {
        var spriteCount = 0
        sprites.forEach {
            let spritePolygon = Polygon(rect: $0.collisionBounds.offsetBy(dx: CGFloat($0.x), dy: CGFloat($0.y)), pivot: CGPoint(x: $0.collisionPivot.x + CGFloat($0.x), y: $0.collisionPivot.y + CGFloat($0.y)), rotation: CGFloat($0.collisionRotation))
            if spritePolygon.intersect(withPolygon: inPolygon) {
                if vector.directionOf(point: CGPoint(x: CGFloat($0.x) + $0.collisionPivot.x, y: CGFloat($0.y) + $0.collisionPivot.y)) >= 0 {
                    spriteCount += 1
                }
            }
        }
        return spriteCount
    }

    
    // --
    // MARK: Configurable values
    // --
    
    var gridWidth: Float = 1 {
        didSet {
            if gridWidth != oldValue {
                UniLayout.setNeedsLayout(view: self)
            }
            physics.width = gridWidth
        }
    }

    var gridHeight: Float = 1 {
        didSet {
            if gridHeight != oldValue {
                UniLayout.setNeedsLayout(view: self)
            }
            physics.height = gridHeight
        }
    }
    
    var sliceWidth: Float = 0

    var fps = 60
    
    var drawPhysicsBoundaries = false

    
    // --
    // MARK: Physics delegate
    // --
    
    func didLethalCollision() {
        physicsDelegate?.didLethalCollision()
    }


    // --
    // MARK: Movement
    // --
    
    private func update(timeInterval: TimeInterval) {
        physics.prepareObjects()
        sprites.forEach {
            $0.update(timeInterval: timeInterval, physics: physics)
        }
        setNeedsDisplay()
    }


    // --
    // MARK: Drawing
    // --
    
    override func draw(_ rect: CGRect) {
        // Draw sprites and optional physics boundaries
        if let context = UIGraphicsGetCurrentContext() {
            let spriteCanvas = SpriteCanvas(context: context, canvasWidth: bounds.width, canvasHeight: bounds.height, gridWidth: CGFloat(gridWidth), gridHeight: CGFloat(gridHeight))
            sprites.forEach {
                $0.draw(canvas: spriteCanvas)
            }
            if drawPhysicsBoundaries {
                for boundary in collisionBoundaries {
                    spriteCanvas.fillRotatedRect(centerX: CGFloat(boundary.x) + boundary.collisionPivot.x, centerY: CGFloat(boundary.y) + boundary.collisionPivot.y, width: CGFloat(boundary.width), height: CGFloat(boundary.height), color: .red, rotation: CGFloat(boundary.collisionRotation))
                }
                if let boundary = sliceVectorBoundary {
                    spriteCanvas.fillRotatedRect(centerX: CGFloat(boundary.x) + boundary.collisionPivot.x, centerY: CGFloat(boundary.y) + boundary.collisionPivot.y, width: CGFloat(boundary.width), height: CGFloat(boundary.height), color: .yellow, rotation: CGFloat(boundary.collisionRotation))
                }
            }
        }

        // Schedule next update
        if !updateScheduled {
            let checkTimeInterval = Date.timeIntervalSinceReferenceDate
            let delayTime = 1.0 / Double(fps) - (checkTimeInterval - lastTimeInterval)
            updateScheduled = true
            DispatchQueue.main.asyncAfter(deadline: .now() + max(0.001, delayTime - timeCorrection), execute: {
                // Try to correct for time lost due to dispatch queue inaccuracy
                let currentTimeInterval = Date.timeIntervalSinceReferenceDate
                if (delayTime >= 0.001) {
                    let lostTimeInterval = (currentTimeInterval - checkTimeInterval) - delayTime
                    if lostTimeInterval < -0.0001 {
                        self.timeCorrection -= 0.0001
                    } else if lostTimeInterval > 0.0001 {
                        self.timeCorrection += 0.0001
                    }
                }
                
                // Continue with the next update
                var difference = currentTimeInterval - self.lastTimeInterval
                self.lastTimeInterval = currentTimeInterval
                self.updateScheduled = false
                if difference > 1.0 / Double(self.fps) * 5 {
                    difference = 1.0 / Double(self.fps) * 5
                }
                self.update(timeInterval: difference)
            })
        }
    }

    
    // --
    // MARK: Custom layout
    // --
    
    override func measuredSize(sizeSpec: CGSize, widthSpec: UniMeasureSpec, heightSpec: UniMeasureSpec) -> CGSize {
        if widthSpec == .limitSize && heightSpec == .limitSize {
            if sizeSpec.width * CGFloat(gridHeight) / CGFloat(gridWidth) <= sizeSpec.height {
                return CGSize(width: sizeSpec.width, height: sizeSpec.width * CGFloat(gridHeight) / CGFloat(gridWidth))
            }
            return CGSize(width: sizeSpec.height * CGFloat(gridWidth) / CGFloat(gridHeight), height: sizeSpec.height)
        } else if widthSpec == .exactSize && heightSpec == .exactSize {
            return CGSize(width: sizeSpec.width, height: sizeSpec.height)
        } else if widthSpec == .limitSize || widthSpec == .exactSize {
            return CGSize(width: sizeSpec.width, height: sizeSpec.width * CGFloat(gridHeight) / CGFloat(gridWidth))
        } else if heightSpec == .limitSize || heightSpec == .exactSize {
            return CGSize(width: sizeSpec.height * CGFloat(gridWidth) / CGFloat(gridHeight), height: sizeSpec.height)
        }
        return super.measuredSize(sizeSpec: sizeSpec, widthSpec: widthSpec, heightSpec: heightSpec)
    }

}
